### PR TITLE
Random Battle Refactor: Check for teammate's weather in script

### PR DESCRIPTION
**Pull request by ascriptmaster**
Refactor randomSet to accept an object with different arguments in it. For
now the object tracks weather and the Mega Pokemon count only, but can be
expanded for Trick Room, etc.

Add some checks in randomSet to accomodate the fact that team weather is
present, as well as some extra rejection guidelines in randomTeam.

----

I spent most of the night working on this when I should have been sleeping,
so it definitely needs to be discussed and polished up. This is a very big
change and definitely could use a lot of improvement. For starters, Pokemon
chosen before the weather is set are unaffected by all the rejection patterns
used after the weather is set (except for Shedinja). Should the rejection for
Pokemon/moves under a certain weather be more lenient? Or should we
attempt to re-build previous sets to accommodate the weather?

@TheImmortal @Joimer @Zarel

### CODE REVIEW

**Comment by Joimer** [comment]
I like this very much, and I think you should attempt to accomodate the current team members with the weather.

**Comment by Zarel** [comment]
What Joim said

File: data/scripts.js [hunk 1]
```diff
@@ -1900,10 +1914,14 @@ exports.BattleScripts = {
...
 		var puCount = 0;
-		var megaCount = 0;
+		var teamDetails = {
+			weather: null,
+			megaCount: 0
+		};
 
 		while (pokemonPool.length && pokemonLeft < 6) {
 			var template = this.getTemplate(this.sampleNoReplace(pokemonPool));
+
```
Comment (Slayer95) [inline data/scripts.js]: ?

File: data/scripts.js [hunk 2]
```diff
@@ -2000,7 +2026,43 @@ exports.BattleScripts = {
...
+			}
+			if (set.moves.indexOf('raindance') > -1 || (set.ability === 'Drizzle' && set.item !== 'Blue Orb')) {
+				if (teamDetails.weather && teamDetails.weather !== 'raindance') continue;
+				teamDetails.weather = 'raindance';
+			}
+			if (set.moves.indexOf('sandstorm') > -1 || set.ability === 'Sand Stream') {
+				if (teamDetails.weather && teamDetails.weather !== 'sandstorm') continue;
+				// Remove Shedinjas from the team since we have sand
+				if (!teamDetails.weather) {
+					for (var i in pokemon) {
```
Comment (Slayer95) [inline data/scripts.js]: Don't use <code>for..in</code> over arrays

File: data/scripts.js (hunk 1 above)
Comment (ascriptmaster) [inline data/scripts.js]: Oh, this was leftover from my debugging. I'll remove it.

File: data/scripts.js [hunk 3]
```diff
@@ -1021,11 +1021,16 @@ exports.BattleScripts = {
...
-	randomSet: function (template, slot, noMega) {
+	randomSet: function (template, slot, teamDetails) {
 		if (slot === undefined) slot = 1;
 		var baseTemplate = (template = this.getTemplate(template));
 		var name = template.name;
 
+		// Legacy / backwards compatability for Seasonals, etc
+		if (typeof teamDetails !== 'object') {
+			teamDetails = {noMega: !!teamDetails};
+		}
```
Comment (Slayer95) [inline data/scripts.js]: Do we really need this?

**Comment by ascriptmaster** [comment]
Made some changes, like the rebuilding of sets.

I'm worried about Pokemon having bad sets under this change, especially when it's a rain or sun setup and fire/water gets rejected based on the weather, but it's hard for me to know which sets will come out bad. (Lilligant comes to mind, though)

**Comment by Joimer** [comment]
You can always give it seeds and loop getting several teams and checking how shit they are :p

**Comment by ascriptmaster** [comment]
I built a mocha test to automate set generations for lower-tier (RU/NU/PU) Pokemon as if it were a Rain team, and the only Pokemon that came up with bad sets in my first run (I only had each Pokemon generate one set for now) were Pyroar, Lilligant, every mono-Fire Pokemon (rejected under rain automatically anyways) and Pokemon that ended up with Sunny Day (which also get rejected). There might be more that I missed, of course.

I think it's best if I took the Pokemon that have the worst sets under sun/rain and hardcode them out as well.

**Comment by ascriptmaster** [comment]
New version. It now rejects Snow Warning if it detects another weather (relevant for Aurorus) and hardcodes Lilligant/Pyroar out.

**Comment by TheImmortal** [comment]
I disagree with almost all of this. We do not build teams in Random Battle. Pokemon are mostly independent and that's how it is going to stay (mentioning Battle Factory as the thing that would be focused on teams). I would prefer if you just implement the feature without making use of it and I will do so.

**Comment by ascriptmaster** [comment]
So only have `teamDetails` available as an option, don't actually implement the weather part?

**Comment by TheImmortal** [comment]
Correct.

**Comment by ascriptmaster** [comment]
If I do that, then all I'm doing is wasting extra cycles creating objects and using access calls, and the change is pointless. One day we can figure out some use for `teamDetails`, but that day is not today.

**Pull request closed without merging**